### PR TITLE
multi-LoRA as extra models in OpenAI server

### LINKED ISSUE #2600: Integrate multi-LoRA functionality with OpenAI server
We should allow LoRAs to be queried using the vLLM OpenAI server.

_Originally posted by @Yard1 in https://github.com/vllm-project/vllm/pull/1804#discussion_r1438452662_

### PR DESCRIPTION
**Pull request by jvmncs**
closes #2600

how to serve the loras (mimicking the [multilora inference example](https://github.com/vllm-project/vllm/blob/main/examples/multilora_inference.py)):
```terminal
$ export LORA_PATH=~/.cache/huggingface/hub/models--yard1--llama-2-7b-sql-lora-test/
$ python -m vllm.entrypoints.api_server \
 --model meta-llama/Llama-2-7b-hf \
 --enable-lora \
 --lora-modules sql-lora=$LORA_PATH sql-lora2=$LORA_PATH
```
the above server will list 3 separate values if the user queries `/models`: one for the base served model, and one each for the specified lora modules. in this case sql-lora and sql-lora2 point to the same underlying lora, but this need not be the case. lora config values take the same values they do in EngineArgs

no work has been done here to scope client permissions to specific models

### CODE REVIEW

**Comment by jvmncs** [comment]
was planning to add a test in `tests/entrypoints/test_openai_server.py`, but it's unclear which loras are usable with the existing zephyr model from that file. happy to change that out with some defaults from huggingface, if that's desirable

**Comment by simon-mo** [comment]
@jvmncs, thanks for the PR. I think you can change the base model to other that works. Any suggestions? If not we can also create a new test file just to test LoRA support.

**Comment by jvmncs** [comment]
this one looks good: https://huggingface.co/typeof/zephyr-7b-beta-lora

File: tests/entrypoints/test_openai_server.py [hunk 1]
```diff
@@ -79,8 +97,25 @@ def client():
...
+
+
+@pytest.mark.parametrize(
+    # first test base model, then test loras
+    "model_name",
+    [MODEL_NAME, "zephyr-lora", "zephyr-lora2"],
+)
+async def test_single_completion(server, client: openai.AsyncOpenAI,
+                                 model_name: str):
+    completion = await client.completions.create(model=model_name,
```
Comment (jvmncs) [inline tests/entrypoints/test_openai_server.py:118]: for some reason this test was failing for all cases when I switched to `MODEL_NAME="mistralai/Mistral-7B-v0.1"`. model was consistently emitting "1999" no matter the prompt/temperature I tried. not really sure why that's the case but reverting to the zephyr model fixed it

**Comment by jvmncs** [comment]
latest commit should be ready for a review, assuming CI passes as it did on my machine

**Comment by jvmncs** [comment]
@simon-mo @Yard1 bumping this for review

**Review by Yard1 — COMMENTED** [review]
This looks pretty good to me, could we add an example of starting an API server with loras?

**Comment by jvmncs** [comment]
> could we add an example of starting an API server with loras

sure, what kind of example were you thinking? other than the command snippet in my [original comment](https://github.com/vllm-project/vllm/pull/2775#issue-2119538768), I'm not sure what that would look like

**Comment by Yard1** [comment]
Hmm, I guess we don't really have an example where we start the server - in that case just extending the docs (https://github.com/vllm-project/vllm/blob/main/docs/source/models/lora.rst) with a snippet of how to run the OpenAI server with lora should be enough!

**Comment by jvmncs** [comment]
@Yard1 should be good to go

**Review by Yard1 — APPROVED** [review]
Thanks, looks good!

**Comment by Wizmak9** [comment]
@Yard1 can we do hot swap of lora like without restarting the base model again and changing lora adapters to that base model on the fly.

**Comment by ajaychinni** [comment]
@Wizmak9 You must initiate your VLLM only using the OpenAI server, following the guidelines provided in the [doumnetation ](https://github.com/vllm-project/vllm/blob/main/docs/source/models/lora.rst)
```
python -m vllm.entrypoints.openai.api_server \
    --model meta-llama/Llama-2-7b-hf \
    --enable-lora \
    --lora-modules sql-lora=~/.cache/huggingface/hub/models--yard1--llama-2-7b-sql-lora-test/ sql-lora2=~/.cache/huggingface/hub/models--yard1--llama-2-7b-sql-lora-test2/
```
In the above command just for example loaded two lora modules `sql-lora` and `sql-lora2`

Now, when making a request, you can specify which LoRA you wish to use. In subsequent requests, you can select different LoRA adapters on the fly, without needing to restart the model.

Python code to simulate the request using `sql-lora`. You can comment out this model and use the base model or another LoRA model in subsequent requests:

```
from openai import OpenAI
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

chat_response = client.chat.completions.create(
    # model =  "meta-llama/Llama-2-7b-hf ",
    model = "sql-lora" ,
    # model = "sql-lora2",
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Write a SQL query to answer the question based on the table schema.\n\n context: CREATE TABLE table_name_74 (icao VARCHAR, airport VARCHAR)\n\n question: Name the ICAO for lilongwe international airport "},
    ]
)
print("Chat response:", chat_response)
```

**Comment by devlup** [comment]
@jvmncs  @xjpang @Yard1  please refer this bug, it seems it only works with one lora https://github.com/vllm-project/vllm/issues/7169

**Pull request merged**